pkg/azure/api/api.go

```diff
@@ -274,32 +288,50 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
 	return vpcs, subnets, nil
 }
 
+func generateIpConfigName() string {
+	return "Cilium-" + rand.RandomRuneWithLen(8)
+}
+
 // AssignPrivateIpAddresses assigns the IPs to the interface as specified by
-// the interfaceID. The provided IPs must belong to the subnet as specified by
-// the subnet ID.
-func (c *Client) AssignPrivateIpAddresses(ctx context.Context, subnetID, interfaceID string, ips []net.IP) error {
-	ipConfigurations := make([]network.InterfaceIPConfiguration, 0, len(ips))
+// the interfaceName.
+func (c *Client) AssignPrivateIpAddresses(ctx context.Context, instanceID, vmssName, subnetID, interfaceName string, addresses int) error {
+	var netIfConfig *compute.VirtualMachineScaleSetNetworkConfiguration
+
+	result, err := c.vmss.Get(ctx, c.resourceGroup, vmssName, instanceID, compute.InstanceView)
+	if err != nil {
+		return fmt.Errorf("failed to get VM %s from VMSS %s: %s", instanceID, vmssName, err)
+	}
 
-	for _, ip := range ips {
-		ipString := ip.String()
+	// Search for the existing network interface configuration
+	if result.NetworkProfileConfiguration != nil {
+		for _, networkInterfaceConfiguration := range *result.NetworkProfileConfiguration.NetworkInterfaceConfigurations {
+			if to.String(networkInterfaceConfiguration.Name) == interfaceName {
+				netIfConfig = &networkInterfaceConfiguration
+				break
+			}
+		}
+	}
+
+	if netIfConfig == nil {
+		return fmt.Errorf("interface %s does not exist in VM %s", interfaceName, instanceID)
+	}
+
+	ipConfigurations := make([]compute.VirtualMachineScaleSetIPConfiguration, 0, addresses)
+	for i := 0; i <= addresses; i++ {
 		ipConfigurations = append(ipConfigurations,
-			network.InterfaceIPConfiguration{
-				InterfaceIPConfigurationPropertiesFormat: &network.InterfaceIPConfigurationPropertiesFormat{
-					PrivateIPAddress: &ipString,
-					Subnet:           &network.Subnet{ID: &subnetID},
+			compute.VirtualMachineScaleSetIPConfiguration{
+				Name: to.StringPtr(generateIpConfigName()),
+				VirtualMachineScaleSetIPConfigurationProperties: &compute.VirtualMachineScaleSetIPConfigurationProperties{
+					PrivateIPAddressVersion: compute.IPv4,
+					Subnet:                  &compute.APIEntityReference{ID: to.StringPtr(subnetID)},
 				},
 			},
 		)
-
 	}
 
-	ifaceParams := network.Interface{
-		ID: &interfaceID,
-		InterfacePropertiesFormat: &network.InterfacePropertiesFormat{
-			IPConfigurations: &ipConfigurations,
-		},
-	}
+	ipConfigurations = append(*netIfConfig.IPConfigurations, ipConfigurations...)
+	netIfConfig.IPConfigurations = &ipConfigurations
 
-	_, err := c.interfaces.CreateOrUpdate(ctx, c.resourceGroup, interfaceID, ifaceParams)
+	_, err = c.vmss.Update(ctx, c.resourceGroup, vmssName, instanceID, result)
```

Comment:
I think we need to wait for this `Update` to complete on the Azure backend otherwise the allocator will kick again at a re-sync and it will double allocate.

`Update` returns a `VirtualMachineScaleSetVMsUpdateFuture` on which you can call https://github.com/Azure/go-autorest/blob/master/autorest/azure/async.go#L144-L152 and that will block until it finishes.

In my testing I was not able to repro the double alloc after I deployed a modified version of your PR.